File: rosetta-test-py/rosetta.py
import lispy
import unittest
import threading
import time
import sys
from pathlib import Path

class TestRosettaUnittestContainer(unittest.TestCase):
    pass


class RosettaFunction(object):
    
    def __init__(self, function, env) -> None:
        self.function = function
        self.env = env
        
    def is_valid(self):
        return self.function is not None

    def __call__(self, *args):
        return self.function(self.env, *args)
    
    def __getitem__(self, key):
        if key == 0:
            return lispy.Sym(self.rosetta_role())
        elif key == 1:
            return self
        else:
            return None 
        
    def rosetta_role(self):
        raise NotImplementedError("RosettaFunction is an abstract class")
        
        
class RosettaSetup(RosettaFunction):
    
    def rosetta_role(self):
        return "setup"
    
    
class RosettaTearDown(RosettaFunction):
    
    def rosetta_role(self):
        return "tearDown"
    
    
class Placeholder(RosettaFunction):

    def __init__(self, name, parameters, doc_string) -> None:
        self.name = name
        self.parameters = parameters
        self.doc_string = doc_string
        super().__init__(None, None)
        
    def rosetta_role(self):
        return "placeholder"
    

def rosetta_assert(value, msg=""):
    assert value, msg
    
def rosetta_assert_eq(expected, actual):
    assert expected == actual, f"{expected} != {actual}"

def rosetta_thread(proc):
    thread = threading.Thread(target=proc)
    thread.daemon = True
    thread.start()
    return thread

def rosetta_thread_kill(thread: threading.Thread):
    pass

def rosetta_thread_join(thread: threading.Thread):
    thread.join()
    
def rosetta_thread_yield():
    time.sleep(0)

class RosettaTestSuite(object):

    # TODO: Consider simplyfing the macros to just add to a single list of annotations and extract the relevant parts afterwards (using functions written in rosetta)
    def __init__(self, bridge_name, file_name) -> None:
        self.lispy_env = lispy.Env(outer=lispy.global_env)
        self.initialize_rosetta_primitives()
        self.initialize_rosetta()
        with open(file_name, "r", encoding="utf-8") as file:
            self.suite_source = file.read()
        
        self.suite = None
        self.bridge_name = bridge_name
        
        self.placeholder_functions = {}
        self.setUp_functions = []
        self.tearDown_functions = []

    def initialize_suite(self):
        """Only called after the Python-side was set up and the suite is ready to run."""
        self.suite = self.eval(self.suite_source)
        self.suite_eval("(suite-set-bridge-name! the_suite bridge_name)", bridge_name=self.bridge_name)

    def suite_name(self):
        return self.suite_eval("(suite-name the_suite)")
    
    def suite_version(self):
        return self.suite_eval("(suite-version the_suite)")
    
    def placeholders(self):
        return self.suite_eval("(suite-placeholders the_suite)")
    
    def root_capability(self):
        return self.suite_eval("(suite-root-capability the_suite)")

    def eval(self, code, env=None):
        if not env:
            env = self.lispy_env
        return lispy.eval(lispy.expand(lispy.parse(code), True), env)
    
    def eval_with_args(self, code, **kwargs):
        env = lispy.Env(outer=self.lispy_env)
        for key, value in kwargs.items():
            env[lispy.Sym(key)] = value
        return self.eval(code, env)
    
    def suite_eval(self, code, **kwargs):
        return self.eval_with_args(code, the_suite=self.suite, **kwargs)
    
    def create_placeholder(self, name, parameters, doc_string=""):
        newPlaceholder = Placeholder(name, parameters, doc_string)
        self.lispy_env[name] = newPlaceholder
        newPlaceholder.env = self.lispy_env
        
        if name in self.placeholder_functions:
            newPlaceholder.function = self.placeholder_functions[name]
        
        return newPlaceholder

    def initialize_rosetta_primitives(self):
        self.lispy_env.update({
            lispy.Sym("create-placeholder"): lambda *args: self.create_placeholder(*args),
            lispy.Sym("is-placeholder?"): lambda x: isinstance(x, Placeholder),
            lispy.Sym("assert"): rosetta_assert,
            lispy.Sym("assert-equal"): rosetta_assert_eq,
            lispy.Sym("is-assertion-error?"): lambda e: isinstance(e, AssertionError),
            lispy.Sym("thread"): rosetta_thread,
            lispy.Sym("thread-wait-for-completion"): rosetta_thread_join,
            lispy.Sym("thread-sleep!"): lambda x: time.sleep(float(x)),
            lispy.Sym("thread-yield"): rosetta_thread_yield,
            lispy.Sym("rosetta-test-host"): lambda: "python"
        })

    def initialize_rosetta(self):
        with open("rosetta-test/rosetta-test.scm", "r") as file:
            self.eval(file.read())

    def placeholder(self, name):
        def decorator(func):
            self.placeholder_functions[lispy.Sym(name)] = func    
            return func
        return decorator
    
    def setUp(self):
        def decorator(func):
            self.setUp_functions.append(func)
            return func
        return decorator
    
    def tearDown(self):
        def decorator(func):
            self.tearDown_functions.append(func)
            return func
        return decorator
    
    def ensure_placeholders_are_valid(self):
        invalid_placeholders = [placeholder for placeholder in self.placeholders() if not placeholder.is_valid()]
        if invalid_placeholders:
            invalid_placeholder_list = "\n".join([("- " + placeholder.name) for placeholder in invalid_placeholders])
            invalid_placeholders_suggestion = "\n\n".join([f"@suite.placeholder(\"{placeholder.name}\")\ndef {placeholder.name.replace('-', '_')}(env,*args):\n\tpass" for placeholder in invalid_placeholders])
            raise Exception(f"Empty placeholders:\n{invalid_placeholder_list}\n\nFix based on:\n{invalid_placeholders_suggestion}")
        
    def install_setUp_tearDown_functions(self):
        for func in self.setUp_functions:
            self.root_capability()[2].insert(0, RosettaSetup(func, self.lispy_env))
        for func in self.tearDown_functions:
            self.root_capability()[3].insert(0, RosettaTearDown(func, self.lispy_env))
    
    def generate_test_name(self, rosetta_test):
        return self.eval_with_args("(test-full-name current_test)", current_test=rosetta_test)
        
    def generate_unittest_test_method(self, rosetta_test):
        return lambda testcase: self.run_test(rosetta_test)
    
    def run_test(self, rosetta_test):
        print(self.eval_with_args("(test-name current_test)", current_test=rosetta_test))
        return self.eval_with_args("(test-run current_test)", current_test=rosetta_test)
    
    def run_unittest(self, only=None, only_capabilities=None, exclude=None, exclude_capabilities=None, expected_failures=[]):
        self.initialize_suite()
        self.install_setUp_tearDown_functions()
        self.ensure_placeholders_are_valid()
        self.lispy_env.update({
            lispy.Sym("root-capability"): self.root_capability(),
        })
        tests = self.eval("(capability-all-tests root-capability)")
        
        test_suite = unittest.TestSuite()
        
        selected_tests = self.eval_with_args(
            "(select-tests the_tests only only_capabilities exclude exclude_capabilities)", 
            the_tests=tests, only=only, only_capabilities=only_capabilities, exclude=exclude, exclude_capabilities=exclude_capabilities)
        
        for test in selected_tests:
            test_name = self.generate_test_name(test)
            setattr(TestRosettaUnittestContainer,
                    test_name,
                    self.generate_unittest_test_method(test))
            test_case = TestRosettaUnittestContainer(test_name)
            if test_name in expected_failures:
                unittest.expectedFailure(test_case)
            test_suite.addTest(test_case)
        unittest.TextTestRunner().run(test_suite)

    def run(self, only=None, only_capabilities=None, exclude=None, exclude_capabilities=None, expected_failures=[]):
        self.initialize_suite()
        self.install_setUp_tearDown_functions()
        self.ensure_placeholders_are_valid()
        self.suite_eval("(suite-set-only-tests! the_suite only)", only=only)
        self.suite_eval("(suite-set-only-capabilities! the_suite only_capabilities)", only_capabilities=only_capabilities)
        self.suite_eval("(suite-set-exclude-tests! the_suite exclude)", exclude=exclude)
        self.suite_eval("(suite-set-exclude-capabilities! the_suite exclude_capabilities)", exclude_capabilities=exclude_capabilities)
        self.suite_eval("(suite-set-expected-failures! the_suite expected_failures)", expected_failures=expected_failures)
        self.suite_eval("(suite-run the_suite argv)", argv=sys.argv)
        

def suite(bridge_name, file_name):
    return RosettaTestSuite(bridge_name, file_name)
    
def fixture_path(file_name: str):
    return (Path('./rosetta-test-suites') / Path(file_name)).absolute().as_posix()

File: rosetta-test-rb/rosetta.rb
require "json"
require "test/unit"
require_relative "scheme"

class RosettaFunction
  attr_accessor :function, :env

  def initialize(func, env)
    @function = func
    @env = env
  end

  def valid?
    !@function.nil?
  end

  def call(*args)
    @function.call(@env, *args)
  end

  def [](key)
    if key == 0
      rosetta_role
    elsif key == 1
      self
    end
  end
end

class RosettaSetup < RosettaFunction
  def rosetta_role
    :setup
  end
end

class RosettaTearDown < RosettaFunction
  def rosetta_role
    :tearDown
  end
end

class Placeholder < RosettaFunction
  attr_accessor :name, :parameters, :doc_string

  def initialize(name, parameters, doc_string)
    super(nil, nil)
    @name = name
    @parameters = parameters
    @doc_string = doc_string
  end

  def [](key)
    :placeholder if key == 0
  end

  def template_string
    params = ["env"]
    params += parameters.map { |param| param.to_s.tr("-", "_") }

    <<~TEMPLATE
      placeholder "#{name}" do |#{params.join(", ")}|
        raise Test::Unit::AssertionFailedError, "Placeholder not implemented: #{@name}"
      end
    TEMPLATE
  end
end

def rosetta_assert(value, msg = "")
  raise Test::Unit::AssertionFailedError, msg || "Expected #{value} to be truthy" unless value
end

def rosetta_assert_eq(expected, actual)
  raise Test::Unit::AssertionFailedError, "Expected #{actual} to be #{expected}" unless expected == actual
end

def rosetta_thread(proc)
  Thread.new { proc.call }
end

def rosetta_thread_kill(thread)
  Thread.kill(thread)
end

def rosetta_thread_join(thread)
  # thread.join
  Thread.kill(thread)
end

def rosetta_thread_sleep(time)
  sleep(time)
end

def rosetta_thread_yield
  # Thread.pass
  sleep(0)
end

class RosettaTestSuite
  attr_reader :scheme_env, :suite_name, :bridge_name, :suite_version, 
    :sources, :placeholders, :root_capability

  def initialize(bridge_name, file_name)
    @scheme_env = Scheme::Environment.new([], [], Scheme::GLOBAL_ENV)
    initialize_rosetta_primitives
    initialize_rosetta
    @suite_source = File.read(file_name)
    @bridge_name = bridge_name

    @suite = nil

    @placeholder_functions = {}
    @set_up_functions = []
    @tear_down_functions = []

    @config = Config.new
  end

  def initialize_suite
    @suite = eval_scheme(@suite_source)
    suite_eval("(suite-set-bridge-name! the_suite bridge_name)", :bridge_name => @bridge_name)
  end

  def eval_scheme(code, env = nil)
    env ||= @scheme_env
    Scheme.evaluate_string(code, env)
  end

  def eval_scheme_with_args(code, **kwargs)
    env = Scheme::Environment.new([], [], @scheme_env)
    kwargs.each do |key, value|
      env[key.to_sym] = value
    end
    eval_scheme(code, env)
  end

  def suite_eval(code, **kwargs)
    eval_scheme_with_args(code, :the_suite=>@suite, **kwargs)
  end

  def suite_placeholders()
    suite_eval("(suite-placeholders the_suite)")
  end

  def suite_root_capability()
    suite_eval("(suite-root-capability the_suite)")
  end

  def create_placeholder(name, parameters, doc_string = "")
    new_placeholder = Placeholder.new(name, parameters, doc_string)
    @scheme_env[name] = new_placeholder
    new_placeholder.env = @scheme_env

    if @placeholder_functions[name]
      new_placeholder.function = @placeholder_functions[name]
    end

    new_placeholder
  end

  def initialize_rosetta_primitives
    primitives = {
      "create-placeholder" => lambda { |*args| create_placeholder(*args) },
      "is-placeholder?" => lambda { |x| x.is_a?(Placeholder) },
      "assert" => method(:rosetta_assert),
      "assert-equal" => method(:rosetta_assert_eq),
      "is-assertion-error?" => lambda { |e| e.is_a?(Test::Unit::AssertionFailedError) },
      "thread" => method(:rosetta_thread),
      "thread-wait-for-completion" => method(:rosetta_thread_join),
      "thread-sleep!" => method(:rosetta_thread_sleep),
      "thread-yield" => method(:rosetta_thread_yield),
      "rosetta-test-host" => lambda { "ruby" }
    }

    primitives.each do |key, value|
      @scheme_env[key.to_sym] = value
    end
  end

  def initialize_rosetta
    rosetta_content = File.read("rosetta-test/rosetta-test.scm")
    eval_scheme(rosetta_content)
  end

  def placeholder(name, &func)
    puts "WARNING: Placeholder #{name} is defined multiple times" if @placeholder_functions.key?(name.to_sym)
    @placeholder_functions[name.to_sym] = func
    func
  end

  def set_up(&func)
    @set_up_functions << func
    func
  end

  def tear_down(&func)
    @tear_down_functions << func
    func
  end

  def ensure_placeholders_are_valid
    invalid_placeholders = suite_placeholders.reject(&:valid?)
    return if invalid_placeholders.empty?

    invalid_placeholder_list = invalid_placeholders.map { |p| "- #{p.name}" }.join("\n")
    placeholders_suggestion = invalid_placeholders.map(&:template_string).join("\n")
    raise [
      "Your test suite is missing definitions for the following placeholders:\n#{invalid_placeholder_list}\n",
      "Implement the missing placeholders:\n#{placeholders_suggestion}"
    ].join("\n")
  end

  def install_set_up_tear_down_functions
    @set_up_functions.each do |func|
      suite_root_capability[2].unshift(RosettaSetup.new(func, @scheme_env))
    end
    @tear_down_functions.each do |func|
      suite_root_capability[3].unshift(RosettaTearDown.new(func, @scheme_env))
    end
  end

  def run
    initialize_suite
    install_set_up_tear_down_functions
    ensure_placeholders_are_valid

    suite_eval("(suite-set-only-tests! the_suite only_tests)", :only_tests => @config.only_tests)
    suite_eval("(suite-set-only-capabilities! the_suite only_capabilities)", :only_capabilities => @config.only_capabilities)
    suite_eval("(suite-set-exclude-tests! the_suite exclude)", :exclude => @config.exclude)
    suite_eval("(suite-set-exclude-capabilities! the_suite exclude_capabilities)", :exclude_capabilities => @config.exclude_capabilities)
    suite_eval("(suite-set-expected-failures! the_suite expected_failures)", :expected_failures => @config.expected_failures)
    suite_eval("(suite-run the_suite argv)", :argv => ARGV)
  end

  CONFIG_FIELDS = [:only_tests, :only_capabilities, :exclude, :exclude_capabilities, :expected_failures]

  # Define a setter for each config field
  CONFIG_FIELDS.each do |field|
    define_method field do |*names|
      @config.public_send(:"#{field}=", names.map(&:to_s))
    end
  end

  class Config
    attr_accessor(*CONFIG_FIELDS)

    def initialize
      @expected_failures = []
    end

    def to_h
      CONFIG_FIELDS.map do |key|
        [key, public_send(key)]
      end.to_h
    end
  end
end

def suite(bridge_name, file_name, &block)
  obj = RosettaTestSuite.new(bridge_name, file_name)
  obj.instance_eval(&block)
  obj.run
end

def fixture_path(file_name)
  File.join(".", "suites", file_name)
end
